haystack/document_stores/elasticsearch_base.py

```diff
@@ -0,0 +1,1696 @@
+from typing import List, Optional, Union, Dict, Any, Generator
+
+import json
+import logging
+import time
+from copy import deepcopy
+from string import Template
+
+import numpy as np
+from scipy.special import expit
+from tqdm.auto import tqdm
+from pydantic.error_wrappers import ValidationError
+
+from haystack.document_stores import KeywordDocumentStore
+from haystack.schema import Document, Label
+from haystack.document_stores.base import get_batches_from_generator
+from haystack.document_stores.filter_utils import LogicalFilterClause
+from haystack.errors import DocumentStoreError, HaystackError
+from haystack.nodes.retriever import DenseRetriever
+
+
+logger = logging.getLogger(__name__)
+
+
+def prepare_hosts(host, port):
+    """
+    Create a list of host(s) + port(s) to allow direct client connections to multiple elasticsearch nodes,
+    in the format expected by the client.
+    """
+    if isinstance(host, list):
+        if isinstance(port, list):
+            if not len(port) == len(host):
+                raise ValueError("Length of list `host` must match length of list `port`")
+            hosts = [{"host": h, "port": p} for h, p in zip(host, port)]
+        else:
+            hosts = [{"host": h, "port": port} for h in host]
+    else:
+        hosts = [{"host": host, "port": port}]
+    return hosts
+
+
+class BaseElasticsearchDocumentStore(KeywordDocumentStore):
+    """
+    Base class implementing the common logic for Elasticsearch and Opensearch
+    """
+
+    def __init__(
+        self,
+        client: Any,
+        index: str = "document",
+        label_index: str = "label",
+        search_fields: Union[str, list] = "content",
+        content_field: str = "content",
+        name_field: str = "name",
+        embedding_field: str = "embedding",
+        embedding_dim: int = 768,
+        custom_mapping: Optional[dict] = None,
+        excluded_meta_data: Optional[list] = None,
+        analyzer: str = "standard",
+        recreate_index: bool = False,
+        create_index: bool = True,
+        refresh_type: str = "wait_for",
+        similarity: str = "dot_product",
+        return_embedding: bool = False,
+        duplicate_documents: str = "overwrite",
+        index_type: str = "flat",
+        scroll: str = "1d",
+        skip_missing_embeddings: bool = True,
+        synonyms: Optional[List] = None,
+        synonym_type: str = "synonym",
+    ):
+        super().__init__()
+
+        self.client = client
+        self._RequestError: Any = Exception
+
+        # configure mappings to ES fields that will be used for querying / displaying results
+        if type(search_fields) == str:
+            search_fields = [search_fields]
+
+        # TODO we should implement a more flexible interal mapping here that simplifies the usage of additional,
```

Comment:
Should we create an issue for this "more flexible inter**n**al mapping" TODO right now? Otherwise it will be forgotten easily.

Reply:
That's not from me, I think we should, yeah

Reply:
@masci I investigated the matter and you can remove the TODO without further work. The background is that we currently store all additional information (all additional fields other than id, content, etc.) that an Elasticsearch search hit contains in the metadata of the Haystack document. 
https://github.com/deepset-ai/haystack/blob/75641dd024976591b96285e6eb7cded6254f87d5/haystack/document_stores/elasticsearch.py#L1367

When the TODO was added in March 2020, this was not the case yet. https://github.com/deepset-ai/haystack/pull/44 For each search hit, only explicitly mentioned fields were stored in the Haystack document.